File: src/com/eagle/programmar/Perl/Statements/Perl_PrintStatement.java
// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 17, 2011

package com.eagle.programmar.Perl.Statements;

import java.util.ArrayList;

import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnable;
import com.eagle.programmar.Perl.Perl_Expression;
import com.eagle.programmar.Perl.Symbols.Perl_Variable_Definition;
import com.eagle.programmar.Perl.Terminals.Perl_KeywordChoice;
import com.eagle.programmar.Perl.Terminals.Perl_Literal;
import com.eagle.tokens.SeparatedList;
import com.eagle.tokens.TokenChooser;
import com.eagle.tokens.TokenSequence;
import com.eagle.tokens.interfaces.AbstractExpression;
import com.eagle.tokens.interfaces.AbstractStatement;
import com.eagle.tokens.interfaces.AbstractType;
import com.eagle.tokens.interfaces.AbstractVariable;
import com.eagle.tokens.punctuation.PunctuationComma;
import com.eagle.transform.EagleGenerator;
import com.eagle.transform.EagleTransformableStatementList;
import com.eagle.transform.EagleTransformer;
import com.eagle.transform.EagleGenerator.TypeEnum;

public class Perl_PrintStatement extends TokenSequence
		implements EagleRunnable, AbstractStatement, EagleTransformableStatementList
{
	public @S(10) Perl_KeywordChoice PRINT = new Perl_KeywordChoice("print", "printf");
	public @S(20) Perl_PrintWhat what;

	public static class Perl_PrintWhat extends TokenChooser
	{
		public @FIRST Perl_PrintRedirectInput XXredirectInput;
		public @CHOICE Perl_PrintNormal XXprintNormal;
	}

	public static class Perl_PrintRedirectInput extends TokenSequence
	{
		public @S(10) Perl_Variable_Definition id;
		public @S(20) Perl_Literal multiline; // With << or <<< to redirect stdin
	}

	public static class Perl_PrintNormal extends TokenSequence
	{
		public @S(10) SeparatedList<Perl_Expression, PunctuationComma> strings;
		public @S(20) @OPT @CURIOUS("Extra comma") PunctuationComma comma;
	}

	@Override
	public void interpret(EagleInterpreter interpreter)
	{
		if (what.getWhich() instanceof Perl_PrintNormal)
		{
			Perl_PrintNormal prt = (Perl_PrintNormal) what.getWhich();
			for (int i = 0; i < prt.strings.getPrimaryCount(); i++)
			{
				Perl_Expression expr = prt.strings.getPrimaryElement(i);
				String item = interpreter.getStrValue(expr);
				System.out.print(item.replace("\\n", "\n"));
			}
		}
	}

	@Override
	public ArrayList<AbstractStatement> transformStatement(EagleTransformer transformer,
			EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
	{
		ArrayList<AbstractStatement> result = new ArrayList<AbstractStatement>();
		if (what.getWhich() instanceof Perl_PrintNormal)
		{
			Perl_PrintNormal prt = (Perl_PrintNormal) what.getWhich();
			for (int i = 0; i < prt.strings.getPrimaryCount(); i++)
			{
				Perl_Expression expr = prt.strings.getPrimaryElement(i);
				AbstractExpression line = transformer.transformExpression(generator, expr);
				AbstractStatement stmt = generator.newPrintStatement1(line, TypeEnum.STRING, false, false, this);
				result.add(stmt);
			}
		}
		return result;
	}
}

File: csrc/com/eagle/programmar/Perl/Statements/Perl_PrintStatement.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

using System;
using System.Collections.Generic;

// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 17, 2011

namespace com.eagle.programmar.Perl.Statements
{

	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnable = com.eagle.interpret.EagleRunnable;
	using Perl_Expression = com.eagle.programmar.Perl.Perl_Expression;
	using Perl_Variable_Definition = com.eagle.programmar.Perl.Symbols.Perl_Variable_Definition;
	using Perl_KeywordChoice = com.eagle.programmar.Perl.Terminals.Perl_KeywordChoice;
	using Perl_Literal = com.eagle.programmar.Perl.Terminals.Perl_Literal;
	using SeparatedList = com.eagle.tokens.SeparatedList;
	using TokenChooser = com.eagle.tokens.TokenChooser;
	using TokenSequence = com.eagle.tokens.TokenSequence;
	using AbstractExpression = com.eagle.tokens.interfaces.AbstractExpression;
	using AbstractStatement = com.eagle.tokens.interfaces.AbstractStatement;
	using AbstractType = com.eagle.tokens.interfaces.AbstractType;
	using AbstractVariable = com.eagle.tokens.interfaces.AbstractVariable;
	using PunctuationComma = com.eagle.tokens.punctuation.PunctuationComma;
	using EagleGenerator = com.eagle.transform.EagleGenerator;
	using EagleTransformableStatementList = com.eagle.transform.EagleTransformableStatementList;
	using EagleTransformer = com.eagle.transform.EagleTransformer;
	using TypeEnum = com.eagle.transform.EagleGenerator.TypeEnum;

	public class Perl_PrintStatement : TokenSequence, EagleRunnable, AbstractStatement, EagleTransformableStatementList
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Perl.Terminals.Perl_KeywordChoice PRINT = new com.eagle.programmar.Perl.Terminals.Perl_KeywordChoice("print", "printf");
		public Perl_KeywordChoice PRINT = new Perl_KeywordChoice("print", "printf");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) Perl_PrintWhat what;
		public Perl_PrintWhat what;

		public class Perl_PrintWhat : TokenChooser
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @FIRST Perl_PrintRedirectInput XXredirectInput;
			public Perl_PrintRedirectInput XXredirectInput;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @CHOICE Perl_PrintNormal XXprintNormal;
			public Perl_PrintNormal XXprintNormal;
		}

		public class Perl_PrintRedirectInput : TokenSequence
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Perl.Symbols.Perl_Variable_Definition id;
			public Perl_Variable_Definition id;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.programmar.Perl.Terminals.Perl_Literal multiline;
			public Perl_Literal multiline; // With << or <<< to redirect stdin
		}

		public class Perl_PrintNormal : TokenSequence
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.tokens.SeparatedList<com.eagle.programmar.Perl.Perl_Expression, com.eagle.tokens.punctuation.PunctuationComma> strings;
			public SeparatedList<Perl_Expression, PunctuationComma> strings;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) @OPT @CURIOUS("Extra comma") com.eagle.tokens.punctuation.PunctuationComma comma;
			public  OPT;
		}

		public override void interpret(EagleInterpreter interpreter)
		{
			if (what.getWhich() is Perl_PrintNormal)
			{
				Perl_PrintNormal prt = (Perl_PrintNormal) what.getWhich();
				for (int i = 0; i < prt.strings.getPrimaryCount(); i++)
				{
					Perl_Expression expr = prt.strings.getPrimaryElement(i);
					string item = interpreter.getStrValue(expr);
					Console.Write(item.Replace("\\n", "\n"));
				}
			}
		}

		public override List<AbstractStatement> transformStatement(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
		{
			List<AbstractStatement> result = new List<AbstractStatement>();
			if (what.getWhich() is Perl_PrintNormal)
			{
				Perl_PrintNormal prt = (Perl_PrintNormal) what.getWhich();
				for (int i = 0; i < prt.strings.getPrimaryCount(); i++)
				{
					Perl_Expression expr = prt.strings.getPrimaryElement(i);
					AbstractExpression line = transformer.transformExpression(generator, expr);
					AbstractStatement stmt = generator.newPrintStatement(line, EagleGenerator.TypeEnum.STRING, false, false, this);
					result.Add(stmt);
				}
			}
			return result;
		}
	}

}
